Add auth module

uncmmath opened the pull request.

uncmmath committed:
Add auth module

KirSanich commented on auth/src/main/kotlin/ru/surf/auth/controller/v1/AuthController.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @PostMapping(value = ["/register"], produces = [MediaType.APPLICATION_JSON_VALUE])
+    fun register(username: String?, password: String?): Any {
```
А можно избавиться от нулабл переменных, если это валидировать на фронте?
Чтобы там поле ввода по умолчанию должно было иметь какие-то данные и тогда не придется здесь и внизу обмазываться let функцией

KirSanich replied:
Ну или сделать дто и валидировать дто отдельно, а это вот эта вложенность let в контроллере смущает

uncmmath replied:
Можно не обмазываться, но тогда не будет понятных сообщений об ошибке, если мы сделаем невалидный запрос. 
Мы просто кидаем кастомное исключение с осмысленным(пока нет) сообщением об ошибке, если какой-то параметр отсутстует. 
В идеале фронт не должен делать заведомо невалидных запросов (мы же его пишем), но это утопия из серии "просто не совершай ошибок". Если в процессе разработки фронта ты забыл при запросе какой-то параметр передать, легче увидеть это в ответе с error'ом, чем смотреть логи спринга, чтобы понять где ты ошибся.
С точки зрения поведения без разницы что там вылетет - NPE или наше кастомное исключение.

uncmmath replied:
все же сделал dto

KirSanich commented on auth/src/main/kotlin/ru/surf/auth/controller/v1/AuthController.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @PostMapping(value = ["/resetPassword"], produces = [MediaType.APPLICATION_JSON_VALUE])
+    fun resetPassword(token: String?, password: String?): Any {
```
Как выше

KirSanich commented on auth/src/main/kotlin/ru/surf/auth/controller/v1/AuthController.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @PostMapping(value = ["/validate"], produces = [MediaType.APPLICATION_JSON_VALUE])
+    fun validate(token: String?): Any {
```
Как выше

KirSanich commented on auth/src/main/kotlin/ru/surf/auth/controller/v1/AuthController.kt:
```diff
@@ -0,0 +1,64 @@
+@RestController
+@RequestMapping("/auth/v1")
+@CacheConfig(cacheNames = ["AuthController"])
+class AuthController(@Autowired private val authService: AuthService) {
```
Тут тоже

uncmmath committed:
some refactoring and exceptions improvements

KirSanich commented on auth/src/main/kotlin/ru/surf/auth/configuration/CorsConfiguration.kt:
```diff
@@ -10,7 +10,10 @@ import org.springframework.web.servlet.config.annotation.WebMvcConfigurer
     override fun addCorsMappings(registry: CorsRegistry) {
-        registry.addMapping("/**")
-                .allowedMethods("*")
+        registry.apply {
```
Лол, тут кода больше стало?

uncmmath replied:
Ага

uncmmath committed:
small refactoring, bug fixes and more exceptions improvements

soup1ck commented on auth/src/main/kotlin/ru/surf/auth/configuration/CacheConfiguration.kt:
```diff
@@ -0,0 +1,29 @@
+
+@Configuration
+@EnableCaching
+class CacheConf : CachingConfigurerSupport() {
```
После аннотированного метода или класса обычно делают пустую строку

uncmmath replied:
посмотрю потом код стайлы котлина

soup1ck commented on auth/src/main/kotlin/ru/surf/auth/configuration/CorsConfiguration.kt:
```diff
@@ -0,0 +1,16 @@
+@Configuration
+@EnableWebMvc
+class CorsConfiguration : WebMvcConfigurer {
+    override fun addCorsMappings(registry: CorsRegistry) {
```
Аналогично

soup1ck commented on auth/src/main/kotlin/ru/surf/auth/exception/handling/ExceptionHandler.kt:
```diff
@@ -0,0 +1,32 @@
+@ControllerAdvice
+class ExceptionHandler {
+    private val logger: Logger = LoggerFactory.getLogger(ExceptionHandler::class.java.name)
+
```
вот тут гуд

uncmmath committed:
enhance build stage, add run configuration and remove secrets from application-prod.yml for auth service

uncmmath committed:
add domain model for auth service, make preparations to integrate auth service with core

KirSanich commented on auth/src/main/kotlin/ru/surf/auth/controller/v1/AuthController.kt:
```diff
@@ -0,0 +1,47 @@
+    @PostMapping(value = ["/resetPassword"], produces = [MediaType.APPLICATION_JSON_VALUE])
+    fun resetPassword(@RequestBody dto: ResetPassphraseDto): Any {
+        authService.resetPassword(dto)
+        @Suppress("unused") return object {
```
Может все же респонс ентити сделать везде вместо этих синглтонов?
Туда код можно пропихнуть, хедеры, прочую шнягу

uncmmath replied:
Ну, Никита вообще предложил кидать просто 200 OK с пустым телом вместо этих "success": "true", можно так и сделать

KirSanich replied:
Да, лучше респонс ентити тогда.

uncmmath replied:
сделал

KirSanich commented on auth/src/main/kotlin/ru/surf/auth/entity/AccountIdentity.kt:
```diff
@@ -0,0 +1,12 @@
+
+
+@Entity
+class AccountIdentity(
```
Надо ли все новые сущности вносить в миграцию?

uncmmath replied:
Это хороший вопрос; по сути эти сущности относятся только к auth сервису, и вовне никуда не уходят, так что не уверен насчет этого

KirSanich replied:
Тогда позже решим.

uncmmath committed:
[first approach] add auth in core with spring security

- authorizing candidates by promotion link (CandidateAuthenticationFilter)
- authorizing accounts by credentials (email/pass) (CredentialsAuthenticationFilter)
- authenticating accounts by access token (TokenAuthenticationFilter)

KirSanich commented on core/src/main/kotlin/ru/surf/core/repository/AccountRepository.kt:
```diff
@@ -2,7 +2,8 @@ package ru.surf.core.repository
+import java.util.*
 
 interface AccountRepository : JpaRepository<Account, UUID> {
+    fun findByEmail(email: String): Optional<Account>
```
Это джавовская штука, можно же elvis юзать

uncmmath replied:
Да, в сервисах я их и юзаю где у нас может быть null
А тут сделал с Optional для единообразия с интерфейсом Jpa, там же у нас findById Optional возвращает вроде

Или ок, если у нас будут вперемешку findBy с Optional и elvis?

KirSanich replied:
findByIdOrNull метод есть у репозиториев, видимо для котлина суппорт.

KirSanich replied:
А, у тебя именованый запрос..

KirSanich commented on core/src/main/kotlin/ru/surf/core/service/impl/CandidateServiceImpl.kt:
```diff
@@ -0,0 +1,96 @@
+
+    @Transactional(propagation = Propagation.REQUIRES_NEW, rollbackFor = [Exception::class])
+    override fun promoteCandidate(candidate: Candidate, candidatePromotionDto: CandidatePromotionDto): Account =
+            // TODO в черновом виде
+            accountRepository.run {
+                save(Account(
+                    email = candidate.email
+                )).apply {
+                    flush()
+                }
+            }.apply {
+                traineeRepository.run {
+                    save(Trainee(
+                            isActive = true,
+                            candidate = candidate,
+                            account = this@apply,
+                            event = candidate.events.last()
+                    )).apply {
+                        flush()
+                    }
+                }
+            }.apply {
+                credentialsService.promoteCandidate(candidate.id, AccountCredentialsDto(
+                        identity = id,
+                        passphrase = candidatePromotionDto.passphrase
+                ))
+            }
```
Это нечитабельно

uncmmath replied:
Поэтому и написал, что в черновом виде
вообще для всех этих сущностей(Trainee, Account) нужны отдельные сервисы, пока просто не для всех сущностей они есть

KirSanich replied:
Я напишу позже, пока по декомпоузу идем, чтобы сразу кучу кода не впихивать.

KirSanich commented on core/src/main/kotlin/ru/surf/core/service/impl/EventServiceImpl.kt:
```diff
@@ -16,18 +16,18 @@ import java.util.*
         val transientEntity = eventMapper.convertFromPostRequestEventDtoToEventEntity(postRequestEventDto).apply {
-            eventInitiator = surfEmployeeService.getSurfEmployee(initiatorId)
-            eventType = eventTypeService.getEventType(eventTypeId)
+            eventInitiator = initiatorId?.let { surfEmployeeService.getSurfEmployee(it) }
+            eventType = eventTypeId?.let { eventTypeService.getEventType(it) }
```
Почему через let, инициатор и тип ивента не могут же быть null

uncmmath replied:
Пока сделал временно nullable, потому что для функциональности авторизации/регистрации нам нужны ивенты, а контроллеров/сервисов для связанных сущностей(инициатор, тип ивента) ещё нет

делать это в PR'e связанным с авторизацией кажется избыточным

KirSanich replied:
Пометь тогда тудушкой, чтобы не забыть потом поменять.

uncmmath replied:
пометил

uncmmath committed:
small refactoring for auth controller
add todo comment for event related entities, which are temporary nullable

uncmmath committed:
Merge branch 'main' into feature/task-53-add-auth-module
